File: _dsv/exec_groupby.py
import argparse
from ._base import _Base
from ._column_slicer import _ColumnSlicer
from .exec_ import exec_

class exec_groupby(_ColumnSlicer, exec_):
    ''' aggregate rows using python '''
    name = None
    parser = argparse.ArgumentParser(parents=[exec_.parent])
    parser.set_defaults(slurp=True)
    parser.add_argument('-x', '--complement', action='store_true', help='exclude, rather than include, field names')
    parser.add_argument('-k', '--fields', action='append', default=[], help='search only on these fields')
    parser.add_argument('script', nargs='+', help='python statements to run')
    group = parser.add_mutually_exclusive_group()
    group.add_argument('-I', '--ignore-errors', action='store_true', help='do not abort on python errors')
    group.add_argument('-E', '--remove-errors', action='store_true', help='remove rows on python errors')

    def __init__(self, opts):
        super().__init__(opts)
        self.groups = {}

    def on_row(self, row, ofs=b'\x00'):
        key = tuple(self.slice(row, self.opts.complement))
        self.groups.setdefault(key, []).append(row)

    def on_eof(self):
        header = None

        for key, group in self.groups.items():

            if header is None:
                if self.header is None:
                    header = []
                else:
                    header = self.header
                    if not self.opts.bytes:
                        header = self.parse_value(self.header)
                header = self.slice(header, default=lambda i: str(i+1))

            if not self.opts.bytes:
                key = self.parse_value(key)

            self.current_key = dict(zip(header, key))
            self.exec_on_all_rows(group, K=self.current_key)
        _Base.on_eof(self)

    def handle_exec_result(self, result, vars, table):
        if self.expr and self.opts.var in vars:
            if not isinstance(result, dict):
                result = {self.opts.script[-1]: result}
            result = {**self.current_key, **result}

        return super().handle_exec_result(result, vars, table)

File: src/subcommands/exec_groupby.rs
use anyhow::Result;
use crate::base;
use bstr::{BString};
use std::collections::{HashMap, hash_map::Entry};
use clap::{Parser, ArgAction};
use super::exec;
use std::ffi::{CStr};
use crate::column_slicer::ColumnSlicer;
use crate::python;

#[derive(Parser, Default)]
struct OtherOpts {
    #[arg(short = 'k', long, help = "sort based only on these fields")]
    fields: Vec<String>,
    #[arg(short = 'x', long, action = ArgAction::SetTrue, help = "exclude, rather than include, field names")]
    complement: bool,
    #[arg(long, action = ArgAction::SetTrue, help = "treat fields as regexes")]
    regex: bool,
}

#[derive(Parser, Default)]
#[command(about = "aggregate rows using python")]
pub struct Opts {
    #[command(flatten)]
    common: exec::CommonOpts,
    #[command(flatten)]
    other: OtherOpts,
}

pub struct Handler {
    inner: exec::Handler,
    opts: OtherOpts,
    postprocess: python::Object,
    default_key: python::Object,
    column_slicer: ColumnSlicer,
    groups: HashMap<Vec<BString>, Vec<Vec<BString>>>,
    header: Option<Vec<BString>>,
}

unsafe impl Send for Handler {}

const POSTPROCESS: &CStr = c"
if not isinstance(result, BaseTable) or (isinstance(result, Proxy) and (result.__is_column__() or result.__is_row__())):
    if not isinstance(result, dict):
        result = {default_key: result}
    result = {**current_key, **result}
";

impl Handler {
    pub fn new(opts: Opts) -> Result<Self> {
        let mut exec_opts = exec::Opts::default();
        exec_opts.common = opts.common;
        let inner = exec::Handler::new(exec_opts)?;
        let column_slicer = ColumnSlicer::new(&opts.other.fields, opts.other.regex);

        let py = inner.py.acquire_gil();
        let postprocess = py.compile_code_cstr(POSTPROCESS, python::StartToken::File).unwrap();
        let default_key = py.to_str(inner.opts.common.script.last().unwrap()).unwrap();

        drop(py);

        Ok(Self{
            inner,
            opts: opts.other,
            postprocess,
            column_slicer,
            default_key,
            header: None,
            groups: HashMap::new(),
        })
    }
}

impl base::Processor for Handler {

    fn on_header(&mut self, _base: &mut base::Base, header: Vec<BString>) -> Result<bool> {
        self.column_slicer.make_header_map(&header);
        self.inner.process_header(&header);
        self.header = Some(header);
        Ok(false)
    }

    fn on_row(&mut self, _base: &mut base::Base, row: Vec<BString>) -> Result<bool> {
        let key = self.column_slicer.slice(&row, self.opts.complement, true);
        match self.groups.entry(key) {
            Entry::Occupied(mut entry) => { entry.get_mut().push(row); },
            Entry::Vacant(entry) => { entry.insert(vec![row]); },
        }
        Ok(false)
    }

    fn on_eof(&mut self, base: &mut base::Base) -> Result<bool> {
        let mut header: Option<Vec<_>> = None;
        for (key, group) in self.groups.iter() {
            let py = self.inner.py.acquire_gil();
            let header = header.get_or_insert_with(|| {
                let header = self.column_slicer.slice_with(
                    self.header.as_ref().unwrap_or(&vec![]),
                    self.opts.complement,
                    Some(|i| format!("{i}").into()),
                );
                header.iter().map(|col| self.inner.bytes_to_py(&py, col.as_ref())).collect()
            });
            let current_key = py.empty_dict().unwrap();
            for (k, v) in header.iter().zip(key) {
                py.dict_set(current_key, *k, self.inner.bytes_to_py(&py, v.as_ref()));
            }
            drop(py);

            let result = self.inner.run_python(group.iter(), &[(c"current_key", current_key)])?;
            let py = self.inner.py.acquire_gil();
            let result = if self.inner.expr && let Some(result) = result {
                py.dict_clear(self.inner.locals);
                py.dict_set_string(self.inner.locals, c"result", result);
                py.dict_set_string(self.inner.locals, c"default_key", self.default_key);
                py.dict_set_string(self.inner.locals, c"current_key", current_key);

                py.exec_code(self.postprocess, self.inner.globals.as_ptr(), self.inner.locals.as_ptr())?;
                py.dict_get_string(self.inner.locals, c"result")
            } else {
                result
            };
            drop(py);

            if let Some(result) = result {
                if self.inner.handle_result(base, result)? {
                    return Ok(true)
                }
            }
        }
        base.on_eof()
    }

}
